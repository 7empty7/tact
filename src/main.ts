import { CompilerContext } from "./ast/context";
import fs from 'fs';
import { resolveTypeDescriptors } from "./types/resolveTypeDescriptors";
import { resolveExpressionTypes } from "./types/resolveExpressionType";
import { writeProgram } from "./generator/writeProgram";

const stdlib = fs.readFileSync(__dirname + '/../stdlib/stdlib.tact', 'utf-8');
const funcstdlib = fs.readFileSync(__dirname + '/../stdlib/stdlib.fc', 'utf-8');

export function compile(src: string) {
    let ctx = CompilerContext.fromSources([stdlib, src]);
    ctx = resolveTypeDescriptors(ctx);
    ctx = resolveExpressionTypes(ctx);
    let output = writeProgram(ctx);
    let cOutput = funcstdlib + '\n' + output;
    return cOutput;
}