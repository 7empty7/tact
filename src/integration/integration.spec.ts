import fs from 'fs';
import { compile } from '../main';

describe('integration', () => {
    let recs = fs.readdirSync(__dirname + "/test/");
    for (let r of recs) {
        it('should resolve expressions for ' + r, () => {
            let code = fs.readFileSync(__dirname + "/test/" + r, 'utf8');
            let res = compile(code);
            expect(res).toMatchSnapshot();
        });
    }
});