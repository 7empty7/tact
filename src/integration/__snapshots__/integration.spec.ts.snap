// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration should resolve expressions for case-1.tact 1`] = `
"int __tact_add(int a, int b) inline {
    return a + b;
}

int __tact_sub(int a, int b) inline {
    return a - b;
}

int __tact_mul(int a, int b) inline {
    return a * b;
}

int __tact_div(int a, int b) inline {
    return a / b;
}

int __tact_eq(int a, int b) inline {
    return a == b;
}

int __tact_neq(int a, int b) inline {
    return a != b;
}

int __tact_lt(int a, int b) inline {
    return a < b;
}

int __tact_lte(int a, int b) inline {
    return a <= b;
}

int __tact_gt(int a, int b) inline {
    return a > b;
}

int __tact_gte(int a, int b) inline {
    return a >= b;
}

forall X -> X __tact_get(tuple x, int i) asm "INDEXVAR";

forall X -> X __tact_set(tuple x, int i, X v) asm "SETINDEXVARQ";
int isZero(int x, int y, int z) {
    int b = __tact_add(x, y);
    b = __tact_add(b, 1);
    return __tact_or(__tact_gt(b, 0), z);
}"
`;
